examples/Efficient Cavity Control with SNAP Gates.py

```diff
@@ -0,0 +1,337 @@
+    evo = apply_blocks(alphas, thetas, initial)
+    return 1 - fidelity(target, evo)[0][0]
+
+
+# # Optimization using Adam -- case in point for `qgrad`
+# 
+# This is where the power of `qgrad` comes in. 
+# Since qgrad's functions used in this notebook -- basis,
+# to_dm, dag, Displace, fidelity -- support JAX, we can 
+# evaluate the gradient of the cost function in one line 
+# using JAX's `grad`. This saves us painstakingly evaluating 
+# the derivative of the cost function analytically.
+# 
+
+
+alphas = jnp.array([1, 0.2, 0.1 - 1j]) # Displace parameters
+theta1, theta2, theta3 = ([0.5],
+                          [0., 0.5],
+                          [0.1, 0., 0., 0.1])
+# NOTE: No input values to JAX differentiable functions should be int
+thetas = jnp.array([pad_thetas(N, p) for p in [theta1, theta2, theta3]])
+
+init_params = [alphas, thetas]
+opt_init, opt_update, get_params = optimizers.adam(step_size=1e-2)
+opt_state = opt_init([alphas, thetas])
+
+
+def step(i, opt_state, opt_update):
+    params = get_params(opt_state)
+    g = grad(cost)(params, initial_state, target_state)
+    return opt_update(i, g, opt_state)
+
+epochs = 150
+pbar = tqdm(range(epochs))
+
+loss_val = []
+params_hist = []
+
+for i in pbar:
+    opt_state = step(i, opt_state, opt_update)
+    params = get_params(opt_state)
+    params_hist.append(params)
+    f = 1 - cost(params, initial_state, target_state)
+    loss_val.append(1 - f)
+    pbar.set_description("Fidelity {}".format(f))
+
+
+
+def display_evolution(parameters, num_plots = 4):
+    """
+    Displays the intermediate states during the learning schedule
+    """
+    show_state(initial_state)
+    
+    diff = epochs // int(num_plots)
+    for i in range(0, epochs, diff):
+        alphas, thetas = parameters[i]
+        x = apply_blocks(alphas, thetas, initial_state)
+        show_state(x)
+    
+    # final parameters after last epoch
+    alphas, thetas = parameters[-1]
+    x = apply_blocks(alphas, thetas, initial_state)
+    show_state(x)
+    
+display_evolution(params_hist)
+
+
+# # Conclusion
+# 
+# We see that starting from a vacuum state $|0 \rangle$, 
+# we efficiently learn the target state 
+# $b_{1} = \frac{\sqrt 3 |3> + |9>}{2}$, as is corroborated
+# by the cost function plot below.
+# 
+# The desired target state's Hinton plot and Wigner function
+# are shown before the learning scedule starts. It can be 
+# seen that the last row above is _almost_ the same as the 
+# target state, implying we learn parameters $\alpha$ and 
+# $\vec \theta$ efficiently so as to go from the vacuum 
+# state to a desired target state in just **three** applications
+# of blocks, $\hat B$. This tutorial is just meant to show the 
+# usage of `qgrad`. Like the authors in the original paper, one
+# may use more sophistacated techniques like gradient clipping,
+# regularized cost function, and greater number of epochs to
+# achieve even better learning.
+# 
+
+
+
+plt.plot(loss_val)
```

Comment:
You could show the loss and the fidelity together in this same plot. Physics folks would like to see the fidelity. You can plot both in the same plot and use a legend to specify which is loss and which is fidelity. Alternatively just plot fidelity.